_posts/2021-01-30-Writing-Better-Code-Cloud-Code-with-Parse-Server-Example.md

```diff
@@ -0,0 +1,191 @@
+comments: true
+author: dblythy
+categories: [Learn,JavaScript,GitHub]
+---
+
+Parse Server is a great, quick way to create an app backend without requiring years of knowledge and time.
+
+There are a few additional steps you can do to ensure that your code is the best it can be, and be assured that your Parse Server always runs as smoothly as possible, even as your Cloud Code continues to grow.
+
+<!-- more -->
+
+Today, I'm going to introduce you to the newest features in the [Parse Server Example](https://github.com/parse-community/parse-server-example) project, which will help you write better code, specifically linting, testing, and coverage.
+
+# Linting
+
+Having well written code can save you a lot of time in the long run. There is nothing worse than looking back at old code you've written and needing to spend hours to wrap your head around it. 
+
+There have been _many_ improvements to Javascript since the initial release of Parse Server. Here are a few of them:
+
+## Using const or let instead of var
+
+In previous versions of our Parse javascript docs, we had many examples of javascript with `var`.
+
+Using `var` can result in variables becoming accidently overridden, causing unwanted bugs, such as:
+
+```javascript
+Parse.Cloud.beforeSave('test', function(request) {
+  var object = request.object;
+  var user = request.user;
+  var query = new Parse.Query('Object');
+  if (user.get('object')) {
+    // accidently redefine 'object' variable
+    var object = await query.first({useMasterKey:true});
+    user.set('object',object);
+  }
+  // you might expect object to be request.object here, but now it is the object from the query
+});
+```
+As of ES6, we have evolved all our example code away from var.
+
+```javascript
+Parse.Cloud.beforeSave('test', function(request) {
+  const object = request.object;
+  const user = request.user;
+  const query = new Parse.Query('Object');
+  if (user.get('object')) {
+    // if you accidently redefine 'object' variable, you will get a syntax error using const.
+    const secondObject = await query.first({useMasterKey:true});
+    user.set('object',secondObject);
+  }
+  // object will always be request.object
+});
+```
+
+## Object Literal Shorthand Syntax
+
+With ES6, you can declare variables much clearer using shorthand syntax.
+
+ES5:
+
+```
+Parse.Cloud.beforeSave('test', function(request) {
+  var object = request.object;
+  var user = request.user;
+  request.log.info(object, user);
+});
+```
+
+ES6 shorthand syntax:
+
+```
+Parse.Cloud.beforeSave('test', ({object,user,log}) => {
+  log.info(object, user);
+});
+```
+
+## For of loops
+
+With ES6, you can write loops using `for (const object of objects)` rather than `for (var i=0; i<objects.length; i++)`
+
+## Promise.all
+
+Unless your code requires series excecution (i.e one task after an other), you should try to leverage `Promise.all` in your cloud functions. This will allow your cloud code to resolve much faster.
```

Comment:
```suggestion
Unless your code requires series excecution (i.e one task after an other), you should try to leverage `Promise.all` in your Cloud Functions. This will allow your Cloud Code to resolve much faster.
```